In [2]:
import seaborn as sns #titanic dataset preloaded
import pandas as pd #dataframe manipulation
from scipy.stats import f_oneway  # statistical testing - scipy # anova test - foneway

In [3]:
df= sns.load_dataset('titanic')
df

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True


In [13]:
df = df[['age', 'pclass']].dropna()

In [14]:
#compare - we have 2 genders and ages and passenger class

df['pclass'].unique()

#is there any significant relation between ages and passenger class
# 1 is categorical and 1 is quantitative data

array([3, 1, 2])

In [15]:
class_1 = df[df['pclass'] == 1]['age']
class_2 = df[df['pclass'] == 2]['age']
class_3 = df[df['pclass'] == 3]['age']

In [16]:
class_1

1      38.0
3      35.0
6      54.0
11     58.0
23     28.0
       ... 
871    47.0
872    33.0
879    56.0
887    19.0
889    26.0
Name: age, Length: 186, dtype: float64

In [17]:
# Null Hypothesis - no connection between ages and passenger class
# Alternate Hypothesis - Yes connection

f_stats, p_value = f_oneway(class_1, class_2, class_3)
f_stats

np.float64(57.44348434067624)

In [18]:
p_value

np.float64(7.487984171957403e-24)

In [19]:
alpha = 0.05

In [20]:
if p_value< alpha:
    print("We will reject the null hypo and there is significant difference in atleast 1 passenger class")
else: 
    print("Accept the null hypothesis - there is significant difference")

We will reject the null hypo and there is significant difference in atleast 1 passenger class


In [21]:
## This proved that all 3 passenger classes had age distributions in passengers

# ANOVA (Analysis of Variance)

## Definition
ANOVA (Analysis of Variance) is a statistical test used to compare the means of **three or more groups** and determine whether at least one group mean is significantly different from the others.

---

## When to Use ANOVA
Use ANOVA when:
- One variable is **categorical** (groups/categories).
- One variable is **quantitative/numerical**.
- There are **3 or more groups**.

### Examples
- Age vs Passenger Class (1st, 2nd, 3rd)
- Salary vs Department
- Marks vs Teaching Method
- Blood Pressure vs Different Medicines

---

## Why Not Multiple t-tests?
If we perform many t-tests, the probability of making a Type I error increases.

ANOVA performs a **single test** to compare all group means simultaneously.

---

## Hypotheses

### Null Hypothesis (H₀)
All population means are equal.

μ₁ = μ₂ = μ₃ = ... = μₖ

### Alternative Hypothesis (H₁)
At least one group mean is different.

---

## Basic Idea
ANOVA compares:

1. **Between-group variation**
   - Difference between group means.

2. **Within-group variation**
   - Variation inside each group.

If the between-group variation is much larger than the within-group variation, the means are likely different.

---

## F Statistic

F = (Between-group variance) / (Within-group variance)

- Large F-value → Means are significantly different.
- Small F-value → Means are similar.

---

## Decision Rule

If:

p-value < α

Reject H₀ → At least one group mean is different.

If:

p-value ≥ α

Fail to reject H₀ → No significant difference between group means.

---

# In This Example

Categorical Variable:
- `pclass` (1st, 2nd, 3rd class)

Numerical Variable:
- `age`

Question:
**Do passengers of different classes have different average ages?**

Results:

- F-statistic = 57.44
- p-value = 7.49 × 10⁻²⁴
- α = 0.05

Since:

7.49 × 10⁻²⁴ < 0.05

Reject H₀.

**Conclusion:** At least one passenger class has a significantly different average age.

---

# Code Explanation (Short)

```python
df = sns.load_dataset('titanic')
```
Loads the Titanic dataset.

```python
df = df[['age', 'pclass']].dropna()
```
Keeps only `age` and `pclass` columns and removes missing values.

```python
df['pclass'].unique()
```
Shows the available classes: `[1, 2, 3]`.

```python
class_1 = df[df['pclass'] == 1]['age']
class_2 = df[df['pclass'] == 2]['age']
class_3 = df[df['pclass'] == 3]['age']
```
Creates three groups containing ages of passengers from each class.

```python
f_stats, p_value = f_oneway(class_1, class_2, class_3)
```
Performs one-way ANOVA.

Returns:
- `f_stats` → F-statistic
- `p_value` → Significance value

```python
if p_value < alpha:
```
Checks whether the difference between group means is statistically significant.